crates/metaslang/bindings/src/graph/resolver.rs

```diff
@@ -182,6 +184,47 @@ impl Resolver {
             Vec::new()
         }
     }
+
+    pub(crate) fn ensure_all_references_resolved<KT: KindTypes + 'static>(
+        &mut self,
+        graph: &ExtendedStackGraph<KT>,
+    ) {
+        if self.definitions.is_some() {
+            return;
+        }
+
+        // Resolve all references
+        for handle in graph.iter_references() {
+            if !self.references.contains_key(&handle)
+                && graph
+                    .get_file_descriptor(handle)
+                    .is_some_and(|file| file.is_user())
+            {
+                let definition_handles = self.resolve_internal(graph, handle, true);
+                self.references.insert(handle, definition_handles);
+            }
+        }
+
+        // Build reverse mapping from definitions to reference handles
+        let mut definitions: HashMap<GraphHandle, Vec<GraphHandle>> = HashMap::new();
+        for (reference, resolved_definitions) in &self.references {
+            for definition in resolved_definitions {
+                if let Some(references) = definitions.get_mut(definition) {
+                    references.push(*reference);
+                } else {
+                    definitions.insert(*definition, vec![*reference]);
+                }
+            }
+        }
+        self.definitions = Some(definitions);
+    }
+
+    pub(crate) fn definition_to_references(&self, handle: GraphHandle) -> Vec<GraphHandle> {
+        let Some(ref definitions) = self.definitions else {
+            unreachable!("All references have been resolved");
+        };
+        definitions.get(&handle).cloned().unwrap_or_default()
```

Comment:
I think it is the opposite?

```suggestion
        self.definitions
            .expect("References have not been resolved yet")
            .get(&handle).cloned().unwrap_or_default()
```

Reply:
You're right, the message should be opposite. I tried using `expect` first but didn't work because it tries to move the value out of the `Option` (I realize now that I could have used `as_ref()`) and copied the message over. In the `expect` case I always read it as what should have happened: if the unwrapping fails in this case, I expected all references to have been resolved, but didn't.

I'll keep the message but switch to `.as_ref().expect("...")...`

Reply:
>  I always read it as what should have happened:

Unfortunately Rust doesn't differentiate in the output messages between `panic!()` / `expect()` / `unreachable!()` / etc... They are all "panic" messages, and the resulting error would contain the string here as-is..